WalletWasabi/Crypto/ZeroKnowledge/Prover.cs

```diff
@@ -0,0 +1,98 @@
+using NBitcoin.Secp256k1;
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using WalletWasabi.Crypto.Groups;
+using WalletWasabi.Crypto.Randomness;
+using WalletWasabi.Helpers;
+
+namespace WalletWasabi.Crypto.ZeroKnowledge
+{
+	public static class Prover
+	{
+		public static KnowledgeOfDiscreteLog CreateProof(Scalar secret, GroupElement publicPoint, GroupElement generator, WasabiRandom? random = null)
+		{
+			var proof = CreateProof(new[] { (secret, generator) }, publicPoint, random);
+
+			return new KnowledgeOfDiscreteLog(proof.Nonce, proof.Responses.First());
+		}
+
+		public static KnowledgeOfRepresentation CreateProof(IEnumerable<(Scalar secret, GroupElement generator)> secretGeneratorPairs, GroupElement publicPoint, WasabiRandom? random = null)
+		{
+			Guard.False($"{nameof(publicPoint)}.{nameof(publicPoint.IsInfinity)}", publicPoint.IsInfinity);
+
+			var nonce = GroupElement.Infinity;
+			var randomScalars = new List<Scalar>();
+			var publicPointSanity = publicPoint;
+			foreach (var (secret, generator) in secretGeneratorPairs)
+			{
+				Guard.False($"{nameof(secret)}.{nameof(secret.IsOverflow)}", secret.IsOverflow);
+				Guard.False($"{nameof(secret)}.{nameof(secret.IsZero)}", secret.IsZero);
+				Guard.False($"{nameof(generator)}.{nameof(generator.IsInfinity)}", generator.IsInfinity);
+				publicPointSanity -= secret * generator;
+
+				var randomScalar = GetNonZeroRandomScalar(random);
+				randomScalars.Add(randomScalar);
+				var randomPoint = randomScalar * generator;
+				nonce += randomPoint;
+			}
+
+			if (publicPointSanity != GroupElement.Infinity)
+			{
+				throw new InvalidOperationException($"{nameof(publicPoint)} was incorrectly constructed.");
+			}
+
+			var generators = secretGeneratorPairs.Select(x => x.generator);
+			var challenge = Challenge.Build(publicPoint, nonce, generators);
```

Comment:
I implemented it exactly in the same way because @nothingmuch suggested this to me and because I saw other people adding extra info here like string constants and all kind of things but now I understand that **it is unnecessary to add the `generators` here** (it doesn't hurt but they are unnecessary).

In this paper [How not to Prove Yourself: Pitfalls of the Fiat-Shamir Heuristic andApplications to Helios](https://img.chainnews.com/paper/e19bf20d12d116133c87ade7f0b67efb.pdf) the researchers show the incorrect way to do it (hashing only the statement) and the correct way to do it: hashing the commitment, nonce in our case, and the statement. See page 6 (_The  strong  Fiat-Shamir  transformation_)

Reply:
I've been thinking a bit more and all the problems that I read about are when the Verifier interacts with the Prover by requesting proofs or by providing elements that the Prover must include in the challenge computation  (I remember something similar in the MuSig btw). But in this case it think hashing more public elements cannot provide any additional security.

Reply:
hashing the generators indeed seems unnecessary to me given that the statement is also hashed, since the statement is defined with regards to some pre-agreed set of generators

i think normally what the string constants represent is a reference to the statament, as well as additional things to prevent any collisions in that reference

Reply:
fwiw in my non-qualified opinion the most sensible approach to handling such transcripts seems to be something like https://merlin.cool/

Reply:
I'm fine with hashing the generators, it's not like the final hash would be larger :) Anyhow, it seems like you doubt yourselves, so let's keep hashing them? Or maybe @seresistvanandras wanna chime in?

Reply:
i'm opposed, again see my bringing up STROBE/Melrin, etc, we shouldn't just improvise that part, but be careful to ensure every statement is comitted in full with proper delimitation (at least type-length-value if not STROBE base), and then specify this along with test vectors. this is a protocol design decision, not an implementation one.

but i really don't care *right now*: i think we should address such details at the end, after we finalize the range proof variant, so i don't think this should block this PR

Reply:
Please decide if you want this PR to be merged: https://github.com/nopara73/WalletWasabi/pull/34

I'd also prefer if @seresistvanandras would participate in these conversations as he's the cryptographer.

Reply:
> hashing the generators indeed seems unnecessary to me given that the statement is also hashed, since the statement is defined with regards to some pre-agreed set of generators

This sentence suggests some confusion to me. The generators _**are part of**_ the statement, since the statement says, that given some generator **G** (or potentially more in case of the Knowledge or Representation), and point **pG**, you know the scalar **p**. Maybe it would be clearer on the source code level if there would be a _Statement_ struct with fields _Generators_ and the _PublicPoint_? Afterwards you just hash the Statement and the Nonce. Btw, this is exactly what we are doing rn, without making the statement explicit with yet another struct.

I don't quite get this fuss about the challenge. IIUC we agreed to do use the strong Fiat-Shamir transformatin, which hashes the statement and the nonces (first message of the prover). Generators and public points are both part of the statement. We implemented correctly the strong Fiat-Shamir transformation, hence I don't see any blocking factor here.

Reply:
yes, the generators are a part f the statement, along with the public and secret inputs. however, the generators are always constant, and so don't need to be treated as variables, and just comitting to the public inputs and generators doesn't capture the relation expressed by the statement. adding a tag e.g. "WabiSabiSerialNumberRepresentation"  or whatever is enough to capture the relation in the statement since it only needs to be unique, and that would be enough to refer to all static parts of the statement (assuming the secret inputs are not free in the statement expression, but bound with an exists quantifier) so only the public inputs would need to be added at that point, the generators would already be implied